File: comparisons/MinsC2Rust/c_algorithm/Output/test_project/tests/test_sortedarray.rs
use test_project::compare_int::{int_compare};
use test_project::sortedarray::{SortedArray,sortedarray_new, sortedarray_insert, sortedarray_free, sortedarray_first_index, sortedarray_remove, sortedarray_index_of, sortedarray_last_index, sortedarray_remove_range, sortedarray_length, sortedarray_get};
pub const TEST_SIZE: usize = 20;pub fn generate_sortedarray_equ<T>(equ_func: fn(&T, &T) -> bool) -> Option<SortedArray<T>> 
where
    T: Clone + Ord + From<i32>,
{
    const TEST_SIZE: usize = 20;
    const TEST_ARRAY: [i32; TEST_SIZE] = [10, 12, 12, 1, 2, 3, 6, 7, 2, 23, 13, 23, 23, 34, 31, 9, 0, 0, 0, 0];

    let mut sortedarray = sortedarray_new(0, equ_func, |a: &T, b: &T| a.cmp(b))?;

    for i in 0..TEST_SIZE {
        let value = T::from(TEST_ARRAY[i]);
        sortedarray_insert(&mut sortedarray, value);
    }

    Some(sortedarray)
}

pub fn free_sorted_ints<T>(sortedarray: Option<Box<SortedArray<T>>>) {
    if let Some(mut sortedarray) = sortedarray {
        for i in 0..sortedarray_length(&sortedarray) {
            if let Some(pi) = sortedarray_get(&sortedarray, i) {
                // In Rust, we don't need to manually free memory as it is managed by the ownership system.
                // The Vec<T> inside SortedArray will automatically deallocate its memory when it goes out of scope.
            }
        }
        sortedarray_free(Some(sortedarray));
    }
}

pub fn generate_sortedarray<T>() -> Option<SortedArray<T>>
where
    T: Clone + Ord + From<i32>,
{
    generate_sortedarray_equ(|a: &T, b: &T| a == b)
}

pub fn check_sorted_prop<T: PartialOrd>(sortedarray: &SortedArray<T>) {
    for i in 1..sortedarray_length(sortedarray) {
        let prev = sortedarray_get(sortedarray, i - 1).unwrap();
        let curr = sortedarray_get(sortedarray, i).unwrap();
        assert!(int_compare(prev, curr) <= 0);
    }
}

#[test]
pub fn test_sortedarray_index_of() {
    const TEST_SIZE: usize = 20;
    let sortedarray = generate_sortedarray::<i32>();

    if let Some(sortedarray) = sortedarray {
        for i in 0..TEST_SIZE {
            if let Some(value) = sortedarray_get(&sortedarray, i) {
                let r = sortedarray_index_of(&sortedarray, *value);
                assert!(r >= 0);
                assert_eq!(*sortedarray_get(&sortedarray, r as usize).unwrap(), *value);
            }
        }

        free_sorted_ints(Some(Box::new(sortedarray)));
    }
}

#[test]
pub fn test_sortedarray_index_of_equ_key() {
    // replace equal function by function which checks pointers
    let sortedarray = generate_sortedarray_equ(ptr_equal as fn(&i32, &i32) -> bool);
    let sortedarray = match sortedarray {
        Some(array) => array,
        None => return,
    };

    // check if all search value return the same index
    for i in 0..TEST_SIZE {
        let value = match sortedarray_get(&sortedarray, i) {
            Some(v) => v.clone(),
            None => continue,
        };
        let r = sortedarray_index_of(&sortedarray, value);
        assert!(r >= 0);
        assert_eq!(i, r as usize);
    }

    free_sorted_ints(Some(Box::new(sortedarray)));
}

pub fn ptr_equal<T>(a: &T, b: &T) -> bool {
    std::ptr::eq(a, b)
}

#[test]
pub fn test_sortedarray_remove_range() {
    const TEST_REMOVE_RANGE: usize = 7;
    const TEST_REMOVE_RANGE_LENGTH: usize = 4;

    let mut sortedarray = generate_sortedarray::<i32>().expect("Failed to generate sorted array");

    let mut new_values = Vec::with_capacity(TEST_REMOVE_RANGE_LENGTH);
    for i in 0..TEST_REMOVE_RANGE_LENGTH {
        let value = *sortedarray_get(&sortedarray, TEST_REMOVE_RANGE + TEST_REMOVE_RANGE_LENGTH + i).unwrap();
        new_values.push(value);
    }

    sortedarray_remove_range(&mut sortedarray, TEST_REMOVE_RANGE, TEST_REMOVE_RANGE_LENGTH);

    for i in 0..TEST_REMOVE_RANGE_LENGTH {
        let value = *sortedarray_get(&sortedarray, TEST_REMOVE_RANGE + i).unwrap();
        assert_eq!(value, new_values[i]);
    }

    check_sorted_prop(&sortedarray);
    free_sorted_ints(Some(Box::new(sortedarray)));
}

#[test]
pub fn test_sortedarray_remove() {
    let mut sortedarray = generate_sortedarray::<i32>().expect("Failed to generate sorted array");

    let test_remove_el = 15;

    let ip = sortedarray_get(&sortedarray, test_remove_el + 1).expect("Failed to get element at index");
    let i = *ip;

    sortedarray_remove(&mut sortedarray, test_remove_el);

    let new_element = sortedarray_get(&sortedarray, test_remove_el).expect("Failed to get element at index after removal");
    assert_eq!(*new_element, i);

    check_sorted_prop(&sortedarray);
    free_sorted_ints(Some(Box::new(sortedarray)));
}

#[test]
pub fn test_sortedarray_get() {
    let mut i: usize;

    let arr = generate_sortedarray::<i32>();

    if let Some(arr) = arr {
        for i in 0..sortedarray_length(&arr) {
            let value1 = sortedarray_get(&arr, i);
            let value2 = sortedarray_get(&arr, i);
            assert!(value1 == value2);
            assert!(value1.unwrap() == value2.unwrap());
        }

        free_sorted_ints(Some(Box::new(arr)));
    }
}

File: c-algorithm/test/test-sortedarray.c
// 识别到的包含指令
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <time.h>
#include "alloc-testing.h"
#include "framework.h"
#include "compare-int.h"
#include "sortedarray.h"

// 识别到的全局变量
_Atomic size_t num_assert = 0;

// 识别到的宏定义
#define assert(expr)                                                           \
  num_assert += 1;                                                             \
  ((void)sizeof((expr) ? 1 : 0), __extension__({                               \
     if (expr)                                                                 \
       ; /* empty */
#define TEST_SIZE 20
#define TEST_ARRAY {10, 12, 12, 1, 2, 3, 6, 7, 2, 23, 13, 23, 23, 34, 31, 9,\
   	21, -2, -12, -4}
#define TEST_REMOVE_EL 15
#define TEST_REMOVE_RANGE 7
#define TEST_REMOVE_RANGE_LENGTH 4

// 识别到的函数定义
// 函数: check_sorted_prop (line 50)
void check_sorted_prop(SortedArray *sortedarray)
{
	unsigned int i;
	for (i = 1; i < sortedarray_length(sortedarray); i++) {
		assert(int_compare(
		                   sortedarray_get(sortedarray, i-1),
						   sortedarray_get(sortedarray, i)) <= 0);
	}
}

// 函数: free_sorted_ints (line 60)
void free_sorted_ints(SortedArray *sortedarray)
{
	unsigned int i;
	for (i = 0; i < sortedarray_length(sortedarray); i++) {
		int *pi = (int*) sortedarray_get(sortedarray, i);
		free(pi);
	}

	sortedarray_free(sortedarray);
}

// 函数: generate_sortedarray_equ (line 71)
SortedArray *generate_sortedarray_equ(SortedArrayEqualFunc equ_func)
{
	/* generate a sorted array of length TEST_SIZE, filled with random 
	   numbers. */
	SortedArray *sortedarray;
	unsigned int i;

	int array[TEST_SIZE] = TEST_ARRAY;

	sortedarray = sortedarray_new(0, equ_func, int_compare);

	for (i = 0; i < TEST_SIZE; ++i) {
		int *pi = malloc(sizeof(int));
		*pi = array[i];
		sortedarray_insert(sortedarray, pi);
	}

	return sortedarray;
}

// 函数: generate_sortedarray (line 91)
SortedArray *generate_sortedarray(void)
{
	return generate_sortedarray_equ(int_equal);
}

// 函数: test_sortedarray_new_free (line 96)
void test_sortedarray_new_free(void)
{
	SortedArray *sortedarray;

	/* test normal */
	sortedarray = sortedarray_new(0, int_equal, int_compare);
	assert(sortedarray != NULL);
	sortedarray_free(sortedarray);

	/* freeing null */
	sortedarray_free(NULL);

	/* low memory */
	alloc_test_set_limit(0);
	sortedarray = sortedarray_new(0, int_equal, int_compare);
	assert(sortedarray == NULL);

	alloc_test_set_limit(-1);
}

// 函数: test_sortedarray_insert (line 116)
void test_sortedarray_insert(void)
{
	SortedArray *sortedarray = generate_sortedarray();
	unsigned int i;

	/* insert a few random numbers, then check if everything is sorted */
	for (i = 0; i < 20; i++) {
		int i = (int) (((float) rand())/((float) RAND_MAX) * 100);
		int *pi = malloc(sizeof(int));
		*pi = i;
		sortedarray_insert(sortedarray, pi);
	}

	check_sorted_prop(sortedarray);
	free_sorted_ints(sortedarray);
}

// 函数: test_sortedarray_remove (line 133)
void test_sortedarray_remove(void)
{
	SortedArray *sortedarray = generate_sortedarray();

	/* remove index 24 */
	int *ip = (int*) sortedarray_get(sortedarray, TEST_REMOVE_EL + 1);
	int i = *ip;
	free((int*) sortedarray_get(sortedarray, TEST_REMOVE_EL));
	sortedarray_remove(sortedarray, TEST_REMOVE_EL);
	assert(*((int*) sortedarray_get(sortedarray, TEST_REMOVE_EL)) == i);

	check_sorted_prop(sortedarray);
	free_sorted_ints(sortedarray);
}

// 函数: test_sortedarray_remove_range (line 148)
void test_sortedarray_remove_range(void)
{
	SortedArray *sortedarray = generate_sortedarray();

	/* get values in test range */
	int new[TEST_REMOVE_RANGE_LENGTH];
	unsigned int i;
	for (i = 0; i < TEST_REMOVE_RANGE_LENGTH; i++) {
		new[i] = *((int*) sortedarray_get(sortedarray, TEST_REMOVE_RANGE + 
		                                    TEST_REMOVE_RANGE_LENGTH + i));
	}
	
	/* free removed elements */
	for (i = 0; i < TEST_REMOVE_RANGE_LENGTH; i++) {
		free((int*) sortedarray_get(sortedarray, TEST_REMOVE_RANGE + i));
	}

	/* remove */
	sortedarray_remove_range(sortedarray, TEST_REMOVE_RANGE, 
			TEST_REMOVE_RANGE_LENGTH);
	
	/* assert */
	for (i = 0; i < TEST_REMOVE_RANGE_LENGTH; i++) {
		assert(*((int*) sortedarray_get(sortedarray, TEST_REMOVE_RANGE + i)) == 
		                                                                new[i]);
	}

	check_sorted_prop(sortedarray);
	free_sorted_ints(sortedarray);
}

// 函数: test_sortedarray_index_of (line 179)
void test_sortedarray_index_of(void) {
	SortedArray *sortedarray = generate_sortedarray();

	unsigned int i;
	for (i = 0; i < TEST_SIZE; i++) {
		int r = sortedarray_index_of(sortedarray, 
		                sortedarray_get(sortedarray, i));
		assert(r >= 0);
		assert(*((int*) sortedarray_get(sortedarray,(unsigned int) r)) == 
		        *((int*) sortedarray_get(sortedarray, i)));
	}
	
	free_sorted_ints(sortedarray);
}

// 函数: ptr_equal (line 194)
static int ptr_equal(SortedArrayValue v1, SortedArrayValue v2) {
	return v1 == v2;
}

// 函数: test_sortedarray_index_of_equ_key (line 198)
void test_sortedarray_index_of_equ_key(void)
{
	/* replace equal function by function which checks pointers */
	SortedArray *sortedarray = generate_sortedarray_equ(ptr_equal);
	unsigned int i;

	/* check if all search value return the same index */
	for (i = 0; i < TEST_SIZE; i++) {
		int r = sortedarray_index_of(sortedarray, 
		                             sortedarray_get(sortedarray, i));
		assert(r >= 0);
		assert(i == (unsigned int) r);
	}

	free_sorted_ints(sortedarray);
}

// 函数: test_sortedarray_get (line 215)
void test_sortedarray_get(void) {
	unsigned int i;

	SortedArray *arr = generate_sortedarray();

	for (i = 0; i < sortedarray_length(arr); i++) {
		assert(sortedarray_get(arr, i) == sortedarray_get(arr, i));
		assert(*((int*) sortedarray_get(arr, i)) == 
		       *((int*) sortedarray_get(arr, i)));
	}

	free_sorted_ints(arr);
}

// 识别到的全局变量
static UnitTestFunction tests[] = {
	test_sortedarray_new_free,
	test_sortedarray_insert,
	test_sortedarray_remove,
	test_sortedarray_remove_range,
	test_sortedarray_index_of,
	test_sortedarray_index_of_equ_key,
	test_sortedarray_get,
	NULL   
};

// 识别到的函数定义
// 函数: main (line 240)
int main(int argc, char *argv[])
{
	run_tests(tests);
  printf("num_assert: %lu\n", num_assert);
	return 0;
}
